metagraph/src/annotation/row_diff_builder.cpp

```diff
@@ -809,12 +815,67 @@ void convert_batch_to_row_diff(const std::string &pred_succ_fprefix,
     auto tmp_file = [&](size_t s, size_t j, size_t chunk) {
         return tmp_path/fmt::format("{}/col_{}_{}/chunk_{}", s / 100, s, j, chunk);
     };
-    auto dump_chunk_to_disk = [&](const std::vector<T> &v,
-                                  size_t s, size_t j, size_t chunk) {
-        assert(std::is_sorted(v.begin(), v.end()) && "all bits in chunks must be sorted");
+    auto dump_chunk_to_disk = [&](std::vector<T> &v, size_t s, size_t j, size_t chunk) {
+        if (!swap_disk)
+            return;
+        std::sort(v.begin(), v.end());
         Encoder<T>::append_block(v, tmp_file(s, j, chunk));
         row_diff_bits[s][j] += v.size();
+        v.resize(0);
+    };
+
+    const uint32_t chunks_open_per_thread
+            = MAX_NUM_FILES_OPEN / std::max((uint32_t)1, num_threads) / (2 + with_values);
+    if (chunks_open_per_thread < 3) {
+        logger->error("Can't merge with less than 3 chunks per thread open. "
+                      "Max num files open: {}. Current number of threads: {}.",
+                      MAX_NUM_FILES_OPEN, num_threads);
+        exit(1);
+    }
+
+    using CallOnes = std::function<void(const std::function<void(uint64_t)> &)>;
+    std::function<CallOnes(size_t, size_t)> call_ones = [&](size_t s, size_t j) {
+        CallOnes caller = [&,s,j](const std::function<void(uint64_t)> &call) {
+            std::vector<std::string> filenames;
+            // skip chunk with fwd bits which have already been counted if stage 1
+            for (uint32_t chunk = compute_row_reduction ? 1 : 0;
+                                chunk < num_chunks[s][j]; ++chunk) {
+                filenames.push_back(tmp_file(s, j, chunk));
+            }
+
+            const bool remove_chunks = true;
+            uint64_t r = 0;
+            elias_fano::merge_files<T>(filenames, [&](T v) {
+                call(utils::get_first(v));
+                if constexpr(with_values) {
+                    assert(v.second && "zero diffs must have been skipped");
+                    values[s][j][r++] = matrix::encode_diff(v.second);
+                }
+            }, remove_chunks, chunks_open_per_thread);
+        };
+        return caller;
     };
+    if (!swap_disk) {
+        call_ones = [&](size_t s, size_t j) {
+            CallOnes caller = [&,s,j](const std::function<void(uint64_t)> &call) {
+                auto &v = set_rows_fwd[s][j];
```

Comment:
it would be nice if this was named something like `set_rows_fwd_s_j` so it would be easier to find how `set_rows_fwd` is used throughout this function

Reply:
same applies for `set_rows_bwd`

Reply:
too long, this block is too technical anyway